jscomp/runtime/js_array.ml

```diff
@@ -25,216 +25,193 @@
-
-external unshiftMany : 'a array -> int = "unshift"
-[@@mel.send.pipe: 'a t as 'this] [@@mel.splice]
-
-(* Accessor functions
-*)
-external append : 'a -> 'this = "concat"
-[@@mel.send.pipe: 'a t as 'this]
-[@@deprecated "append is not type-safe. Use `concat` instead, and see #1884"]
-
-external concat : 'this -> 'this = "concat" [@@mel.send.pipe: 'a t as 'this]
-
-external concatMany : 'this array -> 'this = "concat"
-[@@mel.send.pipe: 'a t as 'this] [@@mel.splice]
+external unshift : 'a t -> value:'a -> int = "unshift" [@@mel.send]
 
-(* TODO: Not available in Node V4  *)
-external includes : 'a -> bool = "includes"
-[@@mel.send.pipe: 'a t as 'this]
-(** ES2016 *)
+external unshiftMany : 'a t -> values:'a array -> int = "unshift"
+[@@mel.send] [@@mel.splice]
 
-external indexOf : 'a -> int = "indexOf" [@@mel.send.pipe: 'a t as 'this]
+(** Accessor functions *)
 
-external indexOfFrom : 'a -> from:int -> int = "indexOf"
-[@@mel.send.pipe: 'a t as 'this]
+external concat : 'a t -> other:'a t -> 'a t = "concat" [@@mel.send]
 
-external join : 'a t -> string = "join"
-[@@mel.send] [@@deprecated "please use joinWith instead"]
+external concatMany : 'a t -> arrays:'a t array -> 'a t = "concat"
+[@@mel.send] [@@mel.splice]
 
-external joinWith : string -> string = "join" [@@mel.send.pipe: 'a t as 'this]
+external includes : 'a t -> value:'a -> bool = "includes"
+[@@mel.send]
+(** ES2015 *)
 
-external lastIndexOf : 'a -> int = "lastIndexOf"
-[@@mel.send.pipe: 'a t as 'this]
+external indexOf : 'a t -> value:'a -> int = "indexOf" [@@mel.send]
 
-external lastIndexOfFrom : 'a -> from:int -> int = "lastIndexOf"
-[@@mel.send.pipe: 'a t as 'this]
+external indexOfFrom : 'a t -> value:'a -> from:int -> int = "indexOf"
+[@@mel.send]
 
-external lastIndexOf_start : 'a -> int = "lastIndexOf"
-[@@mel.send.pipe: 'a t as 'this] [@@deprecated "Please use `lastIndexOf"]
+external join : 'a t -> sep:string -> string = "join" [@@mel.send]
+external lastIndexOf : 'a t -> value:'a -> int = "lastIndexOf" [@@mel.send]
 
-external slice : start:int -> end_:int -> 'this = "slice"
-[@@mel.send.pipe: 'a t as 'this]
+external lastIndexOfFrom : 'a t -> value:'a -> from:int -> int = "lastIndexOf"
+[@@mel.send]
 
-external copy : 'this = "slice" [@@mel.send.pipe: 'a t as 'this]
-
-external slice_copy : unit -> 'this = "slice"
-[@@mel.send.pipe: 'a t as 'this] [@@deprecated "Please use `copy`"]
-
-external sliceFrom : int -> 'this = "slice" [@@mel.send.pipe: 'a t as 'this]
-
-external slice_start : int -> 'this = "slice"
-[@@mel.send.pipe: 'a t as 'this] [@@deprecated "Please use `sliceFrom`"]
-
-external toString : string = "toString" [@@mel.send.pipe: 'a t as 'this]
-
-external toLocaleString : string = "toLocaleString"
-[@@mel.send.pipe: 'a t as 'this]
+external slice : 'a t -> start:int -> end_:int -> 'a t = "slice" [@@mel.send]
+external copy : 'a t -> 'a t = "slice" [@@mel.send]
+external sliceFrom : 'a t -> pos:int -> 'a t = "slice" [@@mel.send]
```

Comment:
If we used `send.pipe` we could consolidate `slice` and `sliceFrom` under a single binding:

```ml
external slice : ?start:int -> ?end_:int -> 'a t = "slice" [@@mel.send.pipe: 'a t]
```

With `send` it requires unit at the end, but also works:

```ml
external slice : 'a t -> ?start:int -> ?end_:int -> unit -> 'a t = "slice" [@@mel.send]
```

At the very least, if we want to keep both `slice` and `sliceFrom`, I'd use `start` in `sliceFrom`, for consistency with `slice` and the [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice#start).

Reply:
This is a good point, I like the suggestion, and I don't think adding an extra unit is too invasive (they're not compiled to extra arguments in JS).

The only question to solve is whether there's a difference between `['foo', 'bar'].join()` and `['foo','bar'].join(undefined)`

Reply:
Looks like for `join` there's no difference, will check `slice` etc.